src/radical/pilot/utils/prof_utils.py

```diff
@@ -23,182 +24,484 @@
+        ldurations = {'name_of_duration': [{ru.EVENT: 'state',
+                                            ru.STATE: s.STATE_NAME},
+                                           {ru.EVENT: 'event_name',
+                                            ru.STATE: None}]}
+        sdurations = {'name_of_duration': [{'STATE': s.STATE_NAME},
+                                           [{'EVENT': 'event_name'},
+                                            {'STATE': s.STATE_NAME}]]}
+        ldurations = {'name_of_duration': [{ru.EVENT: 'state',
+                                            ru.STATE: s.STATE_NAME},
+                                            [{ru.EVENT: 'event_name',
+                                              ru.STATE: None},
+                                             {ru.EVENT: 'state',
+                                              ru.STATE: s.event_name}]}
+        sdurations = {'name_of_duration': [{'STATE': s.STATE_NAME},
+                                           {'MSG': 'message_name'}]}
+        ldurations = {'name_of_duration': [{ru.EVENT: 'state',
+                                            ru.STATE: s.STATE_NAME},
+                                           {ru.EVENT: 'cmd',
+                                            ru.MSG: 'message_name'}]}
+    '''
+
+    ldurations = {}
+
+    for k,v in sdurations.items():
+
+        ldurations[k] = []
+        for ts in v:
+
+            if type(ts) is dict:
+                ldurations[k].append(_expand_sduration(ts))
+
+            if type(ts) is list:
+                lds = []
+                for i in ts:
+                    lds.append(_expand_sduration(i))
+                ldurations[k].append(lds)
+
+    return ldurations
+
+
+# ----------------------------------------------------------------------------
+#
+def _expand_sduration(sduration):
+    '''
+    Expands a duration expressed in short form to its long form.
+
+    Definitions:
+
+    - Short form duration: one dictionary containing a state or event name.
+    - Long form duration: one dictionary containing two keys, one of type
+      `ru.EVENT` and one of type `ru.STATE`. The `ru.EVENT` key has a string
+      value while the `ru.STATE` key has a `s.STATE_NAME` object as its value.
+
+    Args:
+
+        sduration (dict): a duration in short form
+
+    Return:
+
+        lduration (dict): sduration in long form
+
+    Example:
+
+        sduration = {'STATE': s.STATE_NAME}
+        lduration = {ru.EVENT: 'state', ru.STATE: s.STATE_NAME}
+        sduration = {'EVENT': 'event_name'}
+        lduration = {ru.EVENT: 'event_name', ru.STATE: None}
+        sduration = {'MSG': 'mesage_name'}
+        lduration = {ru.EVENT: 'cmd', ru.MSG: 'message_name'}
+    '''
+
+    assert(len(sduration.keys()) == 1), 'expand only one short form duration'
+    assert(list(sduration.keys())[0] == 'STATE' or
+           list(sduration.keys())[0] == 'EVENT' or
+           list(sduration.keys())[0] == 'MSG'), 'unknown timestamp type'
+
```

Comment:
I have one more (the review submit swallowed it):
if you want, you can write the above shorter as:
```py
assert(list(sduration.keys())[0] in ['STATE', 'EVENT', 'MSG']), 'unknown timestamp type'
```
although I would actually suggest to just return mismatches unchanged:
```
if len(sduration.keys() != 1:
    return sduration
if list(sduration.keys())[0] not in ['STATE', 'EVENT', 'MSG']:
    return sduration
```
The method then only works on the special entries and leves the others unchanged - for example, if you want to add `event: update_pushed, state: AGENT_EXECUTING` where the full qualification would be needed.

Reply:
Thanks. Done.

Reply:
I thought a bit more about this and I think it is too permissive. I still want to exit on invalid short-formed durations (use an unknown short-form key) and on invalid arbitrary durations with more than two types of timestamp, e.g., {event: update_pushed, state: AGENT_EXECUTING, event: update_pushed, state: AGENT_EXECUTING}